README.md
## shell-menu is a simplified menu for shell environment


### Description

The main target of the project is to provide an easy to deploy menu to use in
shell mode, for example in case of remote SSH connection, that allows the user
to easily execute a set of command.

This is an example of a shell-menu session running on Linux:
```
Example of a shell-menu

+----------------------+   +-----------------------+   +----------------------+   +----------------------+
| First submenu title  |   | Second submenu title  |   |  Two text info box   |   | Single text info box |
+----------------------+   +-----------------------+   +----------------------+   +----------------------+
| 1) First command     |   | 10) Fourth command    |   | * Lorem ipsum dolor  |   | Lorem ipsum dolor    |
| 2) Second command    |   | 11) Fifth command     |   | sit amet,            |   | sit amet,            |
| 3) Third command     |   +-----------------------+   | consectetur          |   | consectetur          |
+----------------------+                               | adipiscing elit.     |   | adipiscing elit.     |
                                                       | Maecenas a.          |   | Phasellus et cursus  |
                                                       | * Lorem ipsum dolor  |   | ex. Nullam feugiat   |
                                                       | sit amet,            |   | posuere.             |
                                                       | consectetur          |   +----------------------+
                                                       | adipiscing elit.     |
                                                       | Donec egestas urna   |
                                                       | id dolor tincidunt   |
                                                       | tincidunt.           |
                                                       +----------------------+

giuseppe@stretch make your choice [ "0" to exit ] :
```


### Installation

To install shell-menu the only things to do are to clone the git repository,
and execute the script INSTALL.py using the Python interpreter that you want
for normal shell-menu execution.

This is two examples of installation command, the first one using probably Python 2,
and the second one using probably Python 3:
```
python INSTALL.py
```
```
python3 INSTALL.py
```
Interactive installation will just ask you for a target directory, in which
shell-menu will be installed inside a subdirectory "shell-menu".

If the target directory already contains a shell-menu instance, installation
will propose to switch to update mode. In this mode just sources and minor
files will be changed, the configuration files won't change.


### Configuration

The configuration is based on two JSON format files. The first must be located
in a subdirectory called 'cnf' inside the shell-menu.py directory.
The second one can be saved in any directory of the system where the user that
will execute the shell-menu.py has the read grants.

First configuration file is the main one, and the name must be "shell-menu.json".
The user is free to choose a name for the second one.


### Synopsis

If all cofigurations have been done correctly just execute
```markdown
./shell-menu.py
```
Into the directoy where the program has been installed.


### Requirements

First of all it can be used only in an Unix-like operating system
(Linux, Solaris, HP-UX, ecc.).

For Python interpreter, you need at least Python 2.6+ or Python 3.3+.

In repository source the default environment is Python 3, but during the
installation it will be changed according the version of Python interpreter
used for installation script.


### Author

Giuseppe Biolo - giuseppe (dot) biolo (at) gmail (dot) com


### License

This software is licensed under GPL v3 license.
```
Copyright (c) 2017 Giuseppe Biolo

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You can read the full text of the GNU General Public License version 3
[here](http://www.gnu.org/licenses/).
```


CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

---------------------------------

## 2017-09-05
Born of the project

---------------------------------

## 2017-09-08
Main release with the principal functions implemented

---------------------------------

## 2017-09-11

### Added

- Use of the same configuration JSON for many hostnames and users using the '*'
  marker

---------------------------------

## 2017-09-21

### Added

- Project totally redesigned using objects
- Info box with auto-split text and multi text

### Removed

- Customizable text for interface (english is enought)

---------------------------------

## 2017-09-22

### Added

- Interactive installer with Python interpreter recognition

---------------------------------

## 2017-09-27

### Added

- Use of environment variables into commands path of the user configuration
  JSON, indicated as usual with a "$" followed by the variable name


INSTALL.py

"""shell-menu is a simplified menu for shell environment.

Description:
    The main target of the project is to provide an easy to deploy menu to use
    in shell mode, for example in case of remote SSH connection, that allows
    the user to easily execute a set of command.

    The configuration is based on two JSON format files. The first must be
    located in a subdirectory called 'cnf' inside the shell-menu.py directory.
    The second one can be saved in any directory of the system where the user
    that will execute the shell-menu.py has the read grants.

    First configuration file is the main one, and the name must be
    "shell-menu.json". The user is free to choose a name for the second one.

Author:
    Giuseppe Biolo  < giuseppe.biolo@gmail.com > < https://github.com/gbiolo >

License:
    This file is part of shell-menu.

    shell-menu is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    shell-menu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
"""


# Compatibility with Python 2.6+ and Python 3.3+
from __future__ import print_function
from __future__ import with_statement

import sys
import os
import shutil
import re
from subprocess import call


def ask_user(question):
    """Function to ask something to user in interactive mode.

    The only parameter to provide is the question for the user, and the return
    value is the user answer.
    """
    answer = None
    if sys.version_info[0] == 2:
        answer = raw_input(question)
    elif sys.version_info[0] == 3:
        answer = input(question)
    return answer


def copy_file(source_file, destination):
    """Function to copy a given file to a given path.

    The two arguments that must be provided are:
        - source_file : file name, with absolute system path, of the new file
        - destination : absolute path of the target directory
    If it's already present in the destination directory a file with the same
    name of the new one, the old file will be lost
    """
    if not os.path.exists(destination):
        try:
            print("Creating the destination path " + destination)
            os.mkdir(destination)
        except OSError:
            print("Error in creation of destination path " + destination)
            exit()
    filename = re.split("/", source_file).pop()
    if os.path.exists(destination + "/" + filename):
        print("Updating ", end='')
    else:
        print("Copying ", end='')
    print("file {0}/{1}".format(destination, filename))
    try:
        shutil.copyfile(source_file, destination + "/" + filename)
    except OSError:
        print("Error in creation of {0}/{1}".format(destination, filename))
        exit()


# Execute only in interactive mode
if __name__ == "__main__":

    # Actual python interpreter used to execute this installer
    interpreter_path = sys.executable

    # Local absolute path
    local_path = os.getcwd()

    # Ask installation path to the user
    destination_path = ask_user("Full path of the directory where install shell-menu (must exists) : ")
    destination_path = (destination_path + "/shell-menu")

    # Mode flags
    update_mode = False

    # Check if the installation directory exists
    if os.path.isdir(destination_path):
        # Update mode or directory already present
        if os.path.exists(destination_path + "/shell-menu.py"):
            print("Installation path already contains an installation of shell-menu")
            update_ask = ask_user("Do you want to continue in update mode? [Y/n] ")
            if update_ask == "Y" or update_ask == "":
                update_mode = True
            else:
                print("Please remove installation directory and repeat the installation procedure")
                exit()
        else:
            print("Installation path already exists. Please remove and repeat the installation procedure")
    else:
        # Normal installation
        print("Installation path doesn't exists... It will be created")
        try:
            os.mkdir(destination_path)
        except OSError:
            print("Error in install directory creation")
            exit()

    # Copy main source file inserting the right environment
    print("Set " + interpreter_path + " as Python interpreter")
    print("Creating source file " + destination_path + "/shell-menu.py")
    with open(destination_path + "/shell-menu.py", "w") as out_handler:
        print("#! " + interpreter_path, end="\n", file=out_handler)
        with open(local_path + "/src/shell-menu.py", "r") as in_handler:
            for line in in_handler:
                print(line, end='', file=out_handler)
    call(["chmod", "u+x", destination_path + "/shell-menu.py"])

    # Copy other source files
    for src_file in os.listdir(local_path + "/src/shell-menu"):
        if re.search("\.py$", src_file):
            if os.path.isfile(local_path + "/src/shell-menu/" + src_file):
                copy_file(local_path + "/src/shell-menu/" + src_file,
                          destination_path + "/shell-menu")

    # Copy LICENSE, README.md, CHANGELOG
    for text_file in "LICENSE", "README.md", "CHANGELOG.md":
        copy_file(local_path + "/" + text_file, destination_path)

    # Copy configuration files only if not in update mode
    if not update_mode:
        for cnf_file in os.listdir(local_path + "/cnf"):
            if re.search("\.json$", cnf_file):
                if os.path.isfile(local_path + "/cnf/" + cnf_file):
                    copy_file(local_path + "/cnf/" + cnf_file,
                              destination_path + "/cnf")


src/shell-menu/box.py

"""shell-menu is a simplified menu for shell environment.

Description:
    The main target of the project is to provide an easy to deploy menu to use
    in shell mode, for example in case of remote SSH connection, that allows
    the user to easily execute a set of command.

    The configuration is based on two JSON format files. The first must be
    located in a subdirectory called 'cnf' inside the shell-menu.py directory.
    The second one can be saved in any directory of the system where the user
    that will execute the shell-menu.py has the read grants.

    First configuration file is the main one, and the name must be
    "shell-menu.json". The user is free to choose a name for the second one.

Author:
    Giuseppe Biolo  < giuseppe.biolo@gmail.com > < https://github.com/gbiolo >

License:
    This file is part of shell-menu.

    shell-menu is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    shell-menu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
"""


class Box:
    """This class rappresent a generic object "Box".

    A "Box" object may contains a menu o one (or even more) info.
    """

    def __init__(self):
        """Initialization of an empty box.

        The generic attributes are:
            rows  : array containing the rows that compose the box, don't care
                    if it's a menu or an info box, with header and footer
            size  : width of the box expressed in character
            title : title of the box
            index : index used by the method get_row to keep the index of the
                    last row returned to the caller
        """
        self.rows = []
        self.size = 0
        self.title = ""
        self.index = 0

    def format_string(self, unformatted, length, allign="sx"):
        """Function to format a string to a given length and allignment.

        Input parameters are the following:
            unformatted : string to format
            length      : final length of the formatted string; in case of
                          input string longer than the final length, the final
                          string will be the same of the input one
            allign      : allignment of the final string; allowed values:
                              - sx     : left allignment   (default value)
                              - dx     : right allignment
                              - center : centred string
        Return value is the formatted string.
        """
        if len(unformatted) >= length:
            return unformatted
        formatted = ""
        spaces = (length - len(unformatted))
        if allign == "sx":
            formatted = (unformatted + (' ' * spaces))
        elif allign == "dx":
            formatted = ((' ' * spaces) + unformatted)
        elif allign == "center":
            spaces_sx = int(spaces / 2)
            spaces_dx = (int(spaces / 2) + (spaces % 2))
            formatted = ((' ' * spaces_sx) + unformatted + (' ' * spaces_dx))
        return formatted

    def create_header(self):
        """Create the header for the box reporting the name (centered).

        If the header already exists, the whole content of the "rows" array
        will be erased and replaced by a new header
        """
        if len(self.rows) > 0:
            self.rows = []
        self.rows.append("+-" + ('-'*(self.size-3)) + "+")
        self.rows.append("|" + self.format_string(self.title, (self.size-2),
                                                   "center") + "|")
        self.rows.append("+-" + ('-'*(self.size-3)) + "+")

    def get_row(self):
        """Method that return an element (row) from the "rows" array.

        If the index is over the array elements number, the method return a
        string of the same size of a row but composed by only spaces (useful to
        create many levels shell-menu)
        """
        if self.index <= (len(self.rows) - 1):
            self.index += 1
            return self.rows[(self.index - 1)]
        else:
            return None


src/shell-menu/info.py

"""shell-menu is a simplified menu for shell environment.

Description:
    The main target of the project is to provide an easy to deploy menu to use
    in shell mode, for example in case of remote SSH connection, that allows
    the user to easily execute a set of command.

    The configuration is based on two JSON format files. The first must be
    located in a subdirectory called 'cnf' inside the shell-menu.py directory.
    The second one can be saved in any directory of the system where the user
    that will execute the shell-menu.py has the read grants.

    First configuration file is the main one, and the name must be
    "shell-menu.json". The user is free to choose a name for the second one.

Author:
    Giuseppe Biolo  < giuseppe.biolo@gmail.com > < https://github.com/gbiolo >

License:
    This file is part of shell-menu.

    shell-menu is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    shell-menu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
"""


import re

from box import Box


class Info(Box):
    """Class that rappresent an info box, with one or more text messages."""

    def __init__(self, configuration):
        """Method that initialize a new info box object.

        The new object configuration is extracted from a dictionary, passed as
        argument ("configuration" argument).
        """
        # Initialization of the base object
        Box.__init__(self)

        self.title = configuration["title"]

        # Maximum legnth of each line
        length = configuration["width"]
        if length < len(self.title):
            length = len(self.title)
        self.size = (length + 4)
        Box.create_header(self)

        # Generate the rows of the box
        if len(configuration["text"]) > 1:
            for text in configuration["text"]:
                self.split_and_append(text, length, "* ")
        else:
            self.split_and_append(configuration["text"][0], length)

        # Closing info box
        self.rows.append("+-" + ('-'*(length+1)) + "+")

    def split_and_append(self, unsplitted, length, head=""):
        """Method to split the info box message text into multiple lines.

        The method generates the rows to insert in the box "rows" array; all
        final lines must respect the given length.
        The parameter "head" indicates an header for the first line of the
        splitted output.
        For single text info box no header is inserted, instad for multi text
        info box each text starts with a "* " header
        """
        # Temporary variables for line length and line content
        temp_length = 0
        temp_row = head
        # Split the input string in words and reformat the word one by one
        for word in re.split("\s+", unsplitted):
            if temp_row == head:
                temp_row = ("| " + head)
            # Words with length over the line length, will be splitted into
            # sub-words and a '-' will be inserted at the end of each sub-words
            # (except the last one)
            while len(word) > length:
                self.rows.append(self.format_string(temp_row +
                                 word[0:(length-temp_length-2)] + "- |",
                                 (length+4)))
                word = word[(length-temp_length-1):]
                temp_length = 0
            # There is enought space at the end of the line to append
            # the new word
            if (len(word) + temp_length + 1) <= length:
                temp_row += word + " "
                temp_length += (len(word) + 1)
            # No space for the new word, so the line must be finalized and
            # appended to the rows array of the box
            else:
                self.rows.append(self.format_string(temp_row,
                                                    (length+3)) + "|")
                temp_row = "| " + word + " "
                temp_length = (len(word) + 1)
        # Append the last line created to the array (only if not empty)
        if temp_row != "":
            self.rows.append(self.format_string(temp_row, (length+3)) + "|")


src/shell-menu/menu.py

"""shell-menu is a simplified menu for shell environment.

Description:
    The main target of the project is to provide an easy to deploy menu to use
    in shell mode, for example in case of remote SSH connection, that allows
    the user to easily execute a set of command.

    The configuration is based on two JSON format files. The first must be
    located in a subdirectory called 'cnf' inside the shell-menu.py directory.
    The second one can be saved in any directory of the system where the user
    that will execute the shell-menu.py has the read grants.

    First configuration file is the main one, and the name must be
    "shell-menu.json". The user is free to choose a name for the second one.

Author:
    Giuseppe Biolo  < giuseppe.biolo@gmail.com > < https://github.com/gbiolo >

License:
    This file is part of shell-menu.

    shell-menu is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    shell-menu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
"""


import re
import os

from box import Box


class Menu(Box):
    """Class that rappresent a menu box."""

    def __init__(self, configuration):
        """Initialization of the menu box object.

        The initialization add to the box the attribute "links", a dictionary
        containing all commands indicated into the user configuration JSON.
        The dictionary uses as keys the indexes of the commands, and as values
        the commands to execute (expressed as single string)
        """
        # Initialization of the base object
        Box.__init__(self)
        self.title = configuration["title"]
        # Empty dictionary initialization
        self.links = {}
        # External commands index string maximum length
        commands = configuration["commands"]
        num_commands = len(commands)
        index_length = 1
        while num_commands >= 10:
            index_length += 1
            num_commands = int(num_commands / 10)
        # Box size, calculated on length of menu name and on length of all
        # command names
        header_length = len(self.title) + 2
        for command in commands:
            command_length = (len(command["name"]) + index_length + 4)
            if command_length > header_length:
                header_length = command_length
        self.size = (header_length + 3)
        Box.create_header(self)
        # Create all menu rows that will be inserted into the array "rows"
        # Then header indicating the box name will be added by the base class
        # Then closing line will be added by the base class too
        index = configuration["base"]
        for command in commands:
            index_str = self.format_string(str(index), index_length, "dx")
            self.rows.append("| " + self.format_string(index_str + ") " +
                             command["name"] + " ", header_length, "sx") + '|')
            self.links[str(index)] = command["command"]
            index += 1
        self.rows.append("+-" + ('-'*header_length) + "+")

    def get_command(self, index):
        """Method that return an array containing the command to execute.

        The command string, from the JSON, will be splitted by spaces (the
        first element is the program, the following are the arguments).
        If the user used an environment variable into the path of the command
        to execute, the environment variable is replaced by its value before
        the command is splitted and returned.
        If the "index" argument value is not a valid key for the links
        dictionary, the method will return the value "None" (no command with
        the passed index).
        """
        if index in self.links:
            for variable in re.findall("\$[A-Z|_]+", self.links[index]):
                # Remove the dollar symbol from the variable name
                variable = variable[1:]
                # Replace each environment variable used
                if variable in os.environ:
                    self.links[index] = re.sub("\$"+variable,
                                               os.environ[variable],
                                               self.links[index])
            # Split the command and return
            return re.split("\s+", self.links[index])
        return None


src/shell-menu.py

"""shell-menu is a simplified menu for shell environment.

Description:
    The main target of the project is to provide an easy to deploy menu to use
    in shell mode, for example in case of remote SSH connection, that allows
    the user to easily execute a set of command.

    The configuration is based on two JSON format files. The first must be
    located in a subdirectory called 'cnf' inside the shell-menu.py directory.
    The second one can be saved in any directory of the system where the user
    that will execute the shell-menu.py has the read grants.

    First configuration file is the main one, and the name must be
    "shell-menu.json". The user is free to choose a name for the second one.

Author:
    Giuseppe Biolo  < giuseppe.biolo@gmail.com > < https://github.com/gbiolo >

License:
    This file is part of shell-menu.

    shell-menu is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    shell-menu is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
"""


# Compatibility with Python 2.6+ and Python 3.3+
from __future__ import print_function
from __future__ import with_statement

import json
from socket import gethostname
from getpass import getuser
from subprocess import call
import sys
import termios
import os
import re


# Import the sheel-menu libraries
sys.path = ([sys.path[0] + "/shell-menu"] + sys.path)
from menu import Menu
from info import Info

# Execute only in interactive mode
if __name__ == "__main__":

    # Empty boxes collection
    boxes = []

    # Dictionaries containing the configuration JSONs
    main_conf = None
    menu_conf = None

    # Open and load the main configuration JSON
    with open(sys.path[1] + "/cnf/shell-menu.json", "r") as configuration:
        main_conf = json.load(configuration)

    # Check if the user has defined a configuration file for the current
    # hostname
    # If there is no configuration, it try to search the generic configuration,
    # marked by the "*"
    hostname = gethostname()
    if hostname not in main_conf["configurations"]:
        if "*" in main_conf["configurations"]:
            hostname = "*"
        else:
            print("No configuration for the current hostname (" +
                  hostname + ")")
            exit()

    # Check if the user has defined a configuration file for the current user
    # If there is no configuration, it try to search the generic configuration,
    # marked by the "*"
    user = getuser()
    if user not in main_conf["configurations"][hostname]:
        if "*" in main_conf["configurations"][hostname]:
            user = '*'
        else:
            print("No configuration for the current user (" + user + ")")
            exit()

    # Style parameters to default values
    vmargin = 0
    hmargin = 0
    hpadding = 3

    # If the user inserted a style configuration in the main JSON, read the
    # defined values.
    # Undefined style values will remain to the default value.
    if "style" in main_conf:
        if "vmargin" in main_conf["style"]:
            vmargin = main_conf["style"]["vmargin"]
        if "hmargin" in main_conf["style"]:
            hmargin = main_conf["style"]["hmargin"]
        if "hpadding" in main_conf["style"]:
            hpadding = main_conf["style"]["hpadding"]

    # Exit sequence to the default value and read from the configuration if
    # defined by the user.
    # The default exit sequence is the simple "0" (zero) character
    exit_key = "0"
    if "exit_key" in main_conf:
        exit_key = main_conf["exit_key"]

    # Remove environment variable into user configuration JSON and replace with
    # their values
    for variable in re.findall("\$[A-Z|_]+",
                               main_conf["configurations"][hostname][user]):
        # Remove the dollar symbol from the variable name
        variable = variable[1:]
        # Replace each environment variable used
        if variable in os.environ:
            main_conf["configurations"][hostname][user] = re.sub(
                "\$"+variable, os.environ[variable],
                main_conf["configurations"][hostname][user])

    # Open the specific configuration JSON indicated in the main configuration
    with open(main_conf["configurations"][hostname][user]) as configuration:
        menu_conf = json.load(configuration)
        # Add all menu boxes
        for menu in sorted(menu_conf["menu"].keys()):
            boxes.append(Menu(menu_conf["menu"][menu]))
        # Add all info boxes if any
        if "info" in menu_conf:
            for info in sorted(menu_conf["info"].keys()):
                boxes.append(Info(menu_conf["info"][info]))

    # Till the end of the world... or the user insert the exit choice :)
    while True:

        # Reset each box index
        for box in boxes:
            box.index = 0

        # Call to clear screen
        call("clear")

        # Print the menu global title (main configuration JSON)
        print(("\n"*vmargin) + (' '*hmargin) + menu_conf["title"], end="\n\n")

        # Print, line by line, every box (menu box before, info box after)
        completed = 0
        while completed < len(boxes):
            completed = 0
            print((' '*hmargin), end='')
            for box in boxes:
                row = box.get_row()
                if row:
                    print(row, end=(' '*hpadding))
                else:
                    print((' '*box.size), end=(' '*hpadding))
                    completed += 1
            print('', end="\n")

        # Ask the user for the index of the command to execute
        question = ("{0}{1}@{2} make your choice [ \"{3}\" to exit ] : ".
                    format((' '*hmargin), getuser(), gethostname(), exit_key))
        if sys.version_info[0] == 2:
            choice = raw_input(question)
        elif sys.version_info[0] == 3:
            choice = input(question)
        if choice == exit_key:
            call("clear")
            exit()
        else:
            found = 0
            # Check each menu box if there is a command with the inserted index
            for box in boxes:
                if isinstance(box, Menu):
                    command = box.get_command(choice)
                    if command:
                        found = 1
                        call("clear")
                        # Execution of the command
                        call(command)
            # Command not found in any menu box
            if not found:
                print()
                print("{0}\"{1}\" is not a valid choice".
                      format((' '*hmargin), choice), end="\n\n")

            # Print the 'go back' message and wait until the user press the
            # ENTER button to continue
            fd = sys.stdin.fileno()
            old = termios.tcgetattr(fd)
            new = termios.tcgetattr(fd)
            new[3] = new[3] & ~termios.ECHO
            try:
                termios.tcsetattr(fd, termios.TCSADRAIN, new)
                print((' '*hmargin) + "--------------------")
                # For Python 2.6+ the exception of SyntaxError for empty string
                # must be catched and "pass" the error
                try:
                    input((" "*hmargin) +
                          "Press ENTER to return to shell-menu")
                except SyntaxError:
                    pass
            finally:
                termios.tcsetattr(fd, termios.TCSADRAIN, old)
